File: backend/app/schemas/helpdesk.py
from __future__ import annotations

from datetime import datetime
from typing import Optional, List
import uuid

from pydantic import BaseModel, ConfigDict, field_validator

from app.models.helpdesk import TicketStatus, TicketCategory, TicketPriority, TicketSource


class TicketCommentCreate(BaseModel):
    body: str
    is_internal: bool = False


class TicketCommentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:            uuid.UUID
    ticket_id:     uuid.UUID
    body:          str
    is_internal:   bool
    created_by_id: Optional[uuid.UUID]
    created_by_name: Optional[str] = None
    created_at:    datetime


class TicketCreate(BaseModel):
    customer_id:  Optional[uuid.UUID] = None
    category:     TicketCategory = TicketCategory.OTHER
    priority:     TicketPriority = TicketPriority.MEDIUM
    source:       TicketSource   = TicketSource.MANUAL
    subject:      str
    description:  Optional[str]  = None
    lot_id:       Optional[uuid.UUID] = None
    sla_hours:    Optional[int]  = None  # if None, derived from priority


class TicketUpdate(BaseModel):
    category:     Optional[TicketCategory] = None
    priority:     Optional[TicketPriority] = None
    source:       Optional[TicketSource]   = None
    subject:      Optional[str]            = None
    description:  Optional[str]            = None
    lot_id:       Optional[uuid.UUID]      = None
    sla_hours:    Optional[int]            = None


class TicketAssignRequest(BaseModel):
    user_id: uuid.UUID


class TicketResolveRequest(BaseModel):
    resolution_notes: Optional[str] = None


class TicketCloseRequest(BaseModel):
    customer_satisfaction: Optional[int] = None  # 1–5

    @field_validator("customer_satisfaction")
    @classmethod
    def validate_satisfaction(cls, v: Optional[int]) -> Optional[int]:
        if v is not None and v not in range(1, 6):
            raise ValueError("customer_satisfaction must be 1–5")
        return v


class TicketRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:             uuid.UUID
    ticket_no:      str
    customer_id:    Optional[uuid.UUID]
    customer_name:  Optional[str] = None
    customer_code:  Optional[str] = None
    category:       TicketCategory
    priority:       TicketPriority
    status:         TicketStatus
    source:         TicketSource
    subject:        str
    description:    Optional[str]
    lot_id:         Optional[uuid.UUID]
    lot_number:     Optional[str] = None
    sla_hours:      int
    first_response_at: Optional[datetime]
    resolved_at:    Optional[datetime]
    closed_at:      Optional[datetime]
    resolution_notes: Optional[str]
    customer_satisfaction: Optional[int]
    assigned_to_id: Optional[uuid.UUID]
    assigned_to_name: Optional[str] = None
    created_by_id:  Optional[uuid.UUID]
    created_at:     datetime
    sla_breached:   bool = False
    comments:       List[TicketCommentRead] = []


class HelpdeskDashboard(BaseModel):
    total_tickets:       int
    open_count:          int
    in_progress_count:   int
    escalated_count:     int
    resolved_count:      int
    closed_count:        int
    sla_breached_count:  int
    avg_resolution_hours: Optional[float]
    by_category: dict
    by_priority: dict

File: frontend/src/lib/helpdesk.ts
const BASE = "/api/v1/helpdesk";

export type TicketStatus   = "OPEN" | "IN_PROGRESS" | "ESCALATED" | "RESOLVED" | "CLOSED";
export type TicketCategory = "QUALITY" | "DELIVERY" | "BILLING" | "PRODUCT" | "OTHER";
export type TicketPriority = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
export type TicketSource   = "PHONE" | "EMAIL" | "PORTAL" | "WHATSAPP" | "MANUAL";

export interface TicketComment {
  id: string;
  ticket_id: string;
  body: string;
  is_internal: boolean;
  created_by_id: string | null;
  created_by_name: string | null;
  created_at: string;
}

export interface Ticket {
  id: string;
  ticket_no: string;
  customer_id: string | null;
  customer_name: string | null;
  customer_code: string | null;
  category: TicketCategory;
  priority: TicketPriority;
  status: TicketStatus;
  source: TicketSource;
  subject: string;
  description: string | null;
  lot_id: string | null;
  lot_number: string | null;
  sla_hours: number;
  first_response_at: string | null;
  resolved_at: string | null;
  closed_at: string | null;
  resolution_notes: string | null;
  customer_satisfaction: number | null;
  assigned_to_id: string | null;
  assigned_to_name: string | null;
  created_by_id: string | null;
  created_at: string;
  sla_breached: boolean;
  comments: TicketComment[];
}

export interface HelpdeskDashboard {
  total_tickets: number;
  open_count: number;
  in_progress_count: number;
  escalated_count: number;
  resolved_count: number;
  closed_count: number;
  sla_breached_count: number;
  avg_resolution_hours: number | null;
  by_category: Record<TicketCategory, number>;
  by_priority: Record<TicketPriority, number>;
}

export interface TicketCreate {
  customer_id?: string;
  category: TicketCategory;
  priority: TicketPriority;
  source: TicketSource;
  subject: string;
  description?: string;
  lot_id?: string;
  sla_hours?: number;
}

async function req<T>(url: string, opts?: RequestInit): Promise<T> {
  const r = await fetch(url, { credentials: "include", ...opts });
  if (!r.ok) {
    const err = await r.json().catch(() => ({ detail: r.statusText }));
    throw new Error(err.detail ?? r.statusText);
  }
  return r.json();
}

const json = (body: unknown) => ({
  method: "POST",
  headers: { "Content-Type": "application/json" },
  body: JSON.stringify(body),
});

export const helpdeskApi = {
  list: (p?: { status?: TicketStatus; category?: TicketCategory; priority?: TicketPriority }) => {
    const q = new URLSearchParams();
    if (p?.status)   q.set("status", p.status);
    if (p?.category) q.set("category", p.category);
    if (p?.priority) q.set("priority", p.priority);
    return req<Ticket[]>(`${BASE}/?${q}`);
  },
  dashboard: () => req<HelpdeskDashboard>(`${BASE}/dashboard`),
  get: (id: string) => req<Ticket>(`${BASE}/${id}`),
  create: (body: TicketCreate) => req<Ticket>(BASE + "/", json(body)),
  assign: (id: string, user_id: string) => req<Ticket>(`${BASE}/${id}/assign`, json({ user_id })),
  escalate: (id: string) => req<Ticket>(`${BASE}/${id}/escalate`, { method: "POST" }),
  resolve: (id: string, resolution_notes?: string) =>
    req<Ticket>(`${BASE}/${id}/resolve`, json({ resolution_notes })),
  close: (id: string, customer_satisfaction?: number) =>
    req<Ticket>(`${BASE}/${id}/close`, json({ customer_satisfaction })),
  reopen: (id: string) => req<Ticket>(`${BASE}/${id}/reopen`, { method: "POST" }),
  addComment: (id: string, body: string, is_internal: boolean) =>
    req<Ticket>(`${BASE}/${id}/comments`, json({ body, is_internal })),
};

export const STATUS_COLOR: Record<TicketStatus, string> = {
  OPEN:        "bg-blue-100 text-blue-700",
  IN_PROGRESS: "bg-yellow-100 text-yellow-700",
  ESCALATED:   "bg-orange-100 text-orange-700",
  RESOLVED:    "bg-green-100 text-green-700",
  CLOSED:      "bg-gray-100 text-gray-500",
};

export const PRIORITY_COLOR: Record<TicketPriority, string> = {
  LOW:      "bg-gray-100 text-gray-600",
  MEDIUM:   "bg-blue-100 text-blue-600",
  HIGH:     "bg-orange-100 text-orange-700",
  CRITICAL: "bg-red-100 text-red-700",
};

export function fmtDate(d: string | null) {
  if (!d) return "—";
  return new Date(d).toLocaleString("en-KE", { day: "2-digit", month: "short", hour: "2-digit", minute: "2-digit" });
}
